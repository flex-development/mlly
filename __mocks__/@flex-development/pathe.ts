/**
 * @file Mocks - @flex-development/pathe
 * @module mocks/flex-development/pathe
 * @see https://github.com/flex-development/pathe
 */

/**
 * [`@flex-development/pathe`][1] module type.
 *
 * [1]: https://github.com/flex-development/pathe
 */
type Actual = typeof import('@flex-development/pathe')

/**
 * `@flex-development/pathe` module.
 *
 * @const {Actual} actual
 */
const actual: Actual = await vi.importActual<Actual>('@flex-development/pathe')

export default {
  basename: vi.fn(actual.basename),
  changeExt: vi.fn(actual.changeExt),
  dirname: vi.fn(actual.dirname),
  extname: vi.fn(actual.extname),
  format: vi.fn(actual.format),
  isAbsolute: vi.fn(actual.isAbsolute),
  join: vi.fn(actual.join),
  parse: vi.fn(actual.parse),
  relative: vi.fn(actual.relative),
  removeExt: vi.fn(actual.removeExt),
  resolve: vi.fn(actual.resolve)
}
