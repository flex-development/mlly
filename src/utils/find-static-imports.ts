/**
 * @file findStaticImports
 * @module mlly/utils/findStaticImports
 */

import {
  SpecifierSyntaxKind,
  StatementKind,
  StatementSyntaxKind
} from '#src/enums'
import type { StaticImport } from '#src/interfaces'
import getSpecifierKind from '#src/internal/get-specifier-kind'
import validateString from '#src/internal/validate-string'
import type { NodeError } from '@flex-development/errnode'
import { STATIC_IMPORT_REGEX } from '@flex-development/import-regex'
import { split, trim } from '@flex-development/tutils'

/**
 * Finds all static import statements in `code`. Ignores matches in comments.
 *
 * @see {@linkcode StaticImport}
 * @see https://regex101.com/r/wlYQUN
 * @see https://developer.mozilla.org/docs/Web/JavaScript/Reference/Statements/import
 *
 * @param {string} [code=''] - Code to evaluate
 * @return {StaticImport[]} Static import statement objects
 * @throws {NodeError<TypeError>} If `code` is not a string
 */
const findStaticImports = (code: string = ''): StaticImport[] => {
  validateString(code, 'code')

  return [...code.matchAll(STATIC_IMPORT_REGEX)].map(match => {
    const { 0: code = '', index: start = 0, groups = {} } = match
    const { assertion = '', imports = '', specifier = '', type = '' } = groups

    /**
     * Statement syntax kind.
     *
     * @const {StaticImport['syntax']} syntax
     */
    const syntax: StaticImport['syntax'] = imports === ''
      ? StatementSyntaxKind.SIDE_EFFECT
      : imports.startsWith('* as')
      ? StatementSyntaxKind.NAMESPACE
      : imports.startsWith('{')
      ? StatementSyntaxKind.NAMED
      : /^\w+$/.test(imports)
      ? StatementSyntaxKind.DEFAULT
      : /^\w+,\s*{/.test(imports)
      ? StatementSyntaxKind.DEFAULT_WITH_NAMED
      : StatementSyntaxKind.DEFAULT_WITH_NAMESPACE

    return {
      assertion,
      code,
      end: start + code.length,
      imports: syntax === StatementSyntaxKind.SIDE_EFFECT
        ? []
        : syntax === StatementSyntaxKind.NAMED
        ? split(imports.replace(/^{|}$/g, ''), ',')
          .map(trim)
          .filter(i => !!i.length)
        : syntax === StatementSyntaxKind.DEFAULT_WITH_NAMED
        ? split(imports, ',')
          .map(i => trim(i).replace(/^{|}$/g, ''))
          .map(trim)
          .filter(i => !!i.length)
        : syntax === StatementSyntaxKind.DEFAULT_WITH_NAMESPACE
        ? split(imports, ',')
          .map(trim)
          .filter(i => !!i.length)
        : [imports],
      kind: StatementKind.IMPORT,
      specifier,
      specifier_kind: getSpecifierKind(specifier),
      specifier_syntax: SpecifierSyntaxKind.STATIC,
      start,
      syntax,
      type: !!type
    }
  })
}

export default findStaticImports
