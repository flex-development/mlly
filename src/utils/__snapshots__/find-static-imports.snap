// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`unit:utils/findStaticImports > should return StaticImport object array 1`] = `
[
  {
    "assertion": "",
    "code": "import { defineBuildConfig, type Config } from '@flex-development/mkbuild'",
    "end": 74,
    "imports": [
      "defineBuildConfig",
      "type Config",
    ],
    "kind": "import",
    "specifier": "@flex-development/mkbuild",
    "specifier_kind": "bare",
    "specifier_syntax": "static",
    "start": 0,
    "syntax": "named",
    "type": false,
  },
  {
    "assertion": "",
    "code": "import * as pathe from '@flex-development/pathe'",
    "end": 123,
    "imports": [
      "* as pathe",
    ],
    "kind": "import",
    "specifier": "@flex-development/pathe",
    "specifier_kind": "bare",
    "specifier_syntax": "static",
    "start": 75,
    "syntax": "namespace",
    "type": false,
  },
  {
    "assertion": "",
    "code": "import type {
  Join,
  Nullable,
  Opaque,
  Simplify
} from '@flex-development/tutils'",
    "end": 212,
    "imports": [
      "Join",
      "Nullable",
      "Opaque",
      "Simplify",
    ],
    "kind": "import",
    "specifier": "@flex-development/tutils",
    "specifier_kind": "bare",
    "specifier_syntax": "static",
    "start": 124,
    "syntax": "named",
    "type": true,
  },
  {
    "assertion": "",
    "code": "import consola from 'consola'",
    "end": 242,
    "imports": [
      "consola",
    ],
    "kind": "import",
    "specifier": "consola",
    "specifier_kind": "bare",
    "specifier_syntax": "static",
    "start": 213,
    "syntax": "default",
    "type": false,
  },
  {
    "assertion": "",
    "code": "import { export1, export2 as alias2 } from \\"module-name\\"",
    "end": 299,
    "imports": [
      "export1",
      "export2 as alias2",
    ],
    "kind": "import",
    "specifier": "module-name",
    "specifier_kind": "bare",
    "specifier_syntax": "static",
    "start": 243,
    "syntax": "named",
    "type": false,
  },
  {
    "assertion": "",
    "code": "import { 'string name' as alias } from \\"module-name\\"",
    "end": 353,
    "imports": [
      "'string name' as alias",
    ],
    "kind": "import",
    "specifier": "module-name",
    "specifier_kind": "bare",
    "specifier_syntax": "static",
    "start": 301,
    "syntax": "named",
    "type": false,
  },
  {
    "assertion": "",
    "code": "import defaultExport, { export1 } from \\"module-name\\"",
    "end": 407,
    "imports": [
      "defaultExport",
      "export1",
    ],
    "kind": "import",
    "specifier": "module-name",
    "specifier_kind": "bare",
    "specifier_syntax": "static",
    "start": 355,
    "syntax": "default-with-named",
    "type": false,
  },
  {
    "assertion": "",
    "code": "import defaultExport, * as name from \\"module-name\\"",
    "end": 459,
    "imports": [
      "defaultExport",
      "* as name",
    ],
    "kind": "import",
    "specifier": "module-name",
    "specifier_kind": "bare",
    "specifier_syntax": "static",
    "start": 409,
    "syntax": "default-with-namespace",
    "type": false,
  },
  {
    "assertion": "",
    "code": "import color from 'tinyrainbow'",
    "end": 492,
    "imports": [
      "color",
    ],
    "kind": "import",
    "specifier": "tinyrainbow",
    "specifier_kind": "bare",
    "specifier_syntax": "static",
    "start": 461,
    "syntax": "default",
    "type": false,
  },
  {
    "assertion": "",
    "code": "import { \\"h-i\\" as hi } from './hi.js'",
    "end": 530,
    "imports": [
      "\\"h-i\\" as hi",
    ],
    "kind": "import",
    "specifier": "./hi.js",
    "specifier_kind": "relative",
    "specifier_syntax": "static",
    "start": 493,
    "syntax": "named",
    "type": false,
  },
  {
    "assertion": "",
    "code": "import './side-effect.mjs'",
    "end": 557,
    "imports": [],
    "kind": "import",
    "specifier": "./side-effect.mjs",
    "specifier_kind": "relative",
    "specifier_syntax": "static",
    "start": 531,
    "syntax": "side-effect",
    "type": false,
  },
  {
    "assertion": "{ type: 'json' }",
    "code": "import tsconfig from './tsconfig.json' assert { type: 'json' }",
    "end": 620,
    "imports": [
      "tsconfig",
    ],
    "kind": "import",
    "specifier": "./tsconfig.json",
    "specifier_kind": "relative",
    "specifier_syntax": "static",
    "start": 558,
    "syntax": "default",
    "type": false,
  },
]
`;
